### ISSUE TITLE
Bluetooth pairing fails for Android phones with MediaTek chipset (IDFGH-5014)

### ISSUE THREAD

**Issue by rj747** [espressif/esp-idf#6800]
## Environment

- Development Kit:      [ESP32-Wrover-Kit]
- Kit version (for WroverKit/PicoKit/DevKitC): [v1]
- Module or chip used:  [ESP32-WROVER-B]
- IDF version (run ``git describe --tags`` to find it): [v4.1.1 @ 4ebe0d8a2f0fabf577ce03ab9dec976dd3ce3198 ]
- Build System:         [idf.py]
- Compiler version: [(crosstool-NG esp-2020r3) 8.4.0]
- Operating System:     [Linux]
- Using an IDE?: [No]
- Power Supply:         [external 5V]

## Problem Description
The ESP32 Bluetooth cannot pair with certain Android phones that use MediaTek chips that support Bluetooth 5.0. The affected phones include LG Stylo 6, LG Phoenix 5, and Alcatel 3V. Phones/tablets that do not use the MediaTek chipset for Bluetooth will pair with no issues such as the Pixel, Pixel4 (qualcom) and Hytab-plus-10wb1 (xradio). 

When capturing the Bluetooth pairing packets, I noticed that there was LL_REJECT_IND_EXT opcode from the device that was common for all the affected phones which may indicate why the connection was dropped. Also tried ESP-IDF Pre-release v4.3-beta2 with different bluetooth module options but the issues still persists.

### Expected Behavior
ESP32 should pair with the android phones with no issues.

### Actual Behavior
ESP32 attempts to pair but will fail regardless of retries.

### Steps to reproduce

1. Enable ESP and attempt to pair via one of the affected phones. A modified version of the ESP-IDF Gatt Security Server Demo at the link below can be run on the Wrover kit to reproduce.

### Code to reproduce this issue
https://gist.github.com/rj747/3c686faffea20d258875413165afe8ef

## Debug Logs
https://gist.github.com/rj747/3c686faffea20d258875413165afe8ef
```
D (1391) BT_HCI: HCI Enqueue Comamnd opcode=0x2008

D (1401) BT_HCI: Receive packet event_code=0xe

D (1401) BT_HCI: btu_hcif_command_complete_evt

D (1401) BT_HCI: HCI Enqueue Comamnd opcode=0x2009

D (1411) BT_HCI: Receive packet event_code=0xe

D (1411) BT_HCI: btu_hcif_command_complete_evt

D (1421) BT_SMP: static passkey 123456
D (1421) BT_HCI: HCI Enqueue Comamnd opcode=0x2005

D (1431) BT_HCI: HCI Enqueue Comamnd opcode=0x200a

D (1431) BT_HCI: Receive packet event_code=0xe

D (1441) BT_HCI: btu_hcif_command_complete_evt

D (1441) BT_HCI: Receive packet event_code=0xe

D (1451) BT_HCI: btu_hcif_command_complete_evt

D (1451) BT_HCI: HCI Enqueue Comamnd opcode=0x2006

D (1461) BT_HCI: Receive packet event_code=0xe

D (1461) BT_HCI: btu_hcif_command_complete_evt

D (1471) BT_HCI: HCI Enqueue Comamnd opcode=0x200a

D (1471) BT_HCI: Receive packet event_code=0xe

D (1471) BT_HCI: btu_hcif_command_complete_evt

D (1481) BT_HCI: HCI Enqueue Comamnd opcode=0x200a

D (1491) BT_HCI: Receive packet event_code=0xe

D (1491) BT_HCI: btu_hcif_command_complete_evt

I (1491) SEC_GATTS_DEMO: advertising start success
D (10731) BT_HCI: Receive packet event_code=0x3e

D (10731) BT_HCI: BLE HCI(id=62) event = 0x0a)
D (10731) BT_HCI: btu_free_timer Unable to find expected alarm in hashmap
D (10741) BT_HCI: btu_free_timer Unable to find expected alarm in hashmap
D (10741) BT_HCI: btu_free_timer Unable to find expected alarm in hashmap
D (10751) BT_L2CAP: l2c_ble_link_adjust_allocation  num_hipri: 0  num_lowpri: 1  low_quota: 10  round_robin_quota: 0  qq: 10
D (10761) BT_L2CAP: l2c_ble_link_adjust_allocation LCB 0   Priority: 0  XmitQuota: 10
D (10771) BT_L2CAP:         SentNotAcked: 0  RRUnacked: 0
D (10781) BT_L2CAP: l2cu_allocate_ccb: cid 0x0000
D (10781) BT_HCI: btu_free_quick_timer Unable to find expected alarm in hashmap
D (10791) BT_HCI: btu_free_quick_timer Unable to find expected alarm in hashmap
D (10801) BT_HCI: btu_free_timer Unable to find expected alarm in hashmap
D (10801) BT_L2CAP: l2c_link_adjust_chnl_allocation
D (10811) BT_L2CAP: CID:0x0040 FCR Mode:0 Priority:2 TxDataRate:1 RxDataRate:1 Quota:20
D (10821) BT_HCI: HCI Enqueue Comamnd opcode=0x41d

D (10821) BT_HCI: Receive packet event_code=0xf

D (10881) BT_HCI: Receive packet event_code=0xc

D (10881) BT_HCI: HCI Enqueue Comamnd opcode=0x2016

D (10891) BT_HCI: Receive packet event_code=0xf

D (10891) BT_HCI: Receive packet event_code=0x3e

D (10891) BT_HCI: BLE HCI(id=62) event = 0x04)
D (10901) BT_HCI: HCI Enqueue Comamnd opcode=0x2022

D (10901) BT_L2CAP: l2cu_process_fixed_chnl_resp
D (10901) BT_HCI: Receive packet event_code=0xe

D (10911) BT_HCI: btu_hcif_command_complete_evt

D (10921) BT_SMP: SMDBG l2c smp_connect_callback

I (10921) BT_L2CAP: L2CA_SetDesireRole() new:x1, disallow_switch:0
I (10931) SEC_GATTS_DEMO: ESP_GATTS_CONNECT_EVT
D (10931) BT_SMP: SMP_Pair state=0 br_state=0 flag=0x0

I (10941) BT_L2CAP: L2CA_ConnectFixedChnl() CID: 0x0006  BDA: 62a27aca9037
D (10951) BT_L2CAP: l2cu_allocate_ccb: cid 0x0000
D (10951) BT_HCI: btu_free_quick_timer Unable to find expected alarm in hashmap
D (10961) BT_HCI: btu_free_quick_timer Unable to find expected alarm in hashmap
D (10961) BT_HCI: btu_free_timer Unable to find expected alarm in hashmap
D (10971) BT_L2CAP: l2c_link_adjust_chnl_allocation
D (10981) BT_L2CAP: CID:0x0004 FCR Mode:0 Priority:2 TxDataRate:1 RxDataRate:1 Quota:20
D (10981) BT_HCI: HCI Enqueue Comamnd opcode=0xc35

D (10991) BT_HCI: reassemble_and_dispatch

D (10991) BT_L2CAP: CID:0x0041 FCR Mode:0 Priority:2 TxDataRate:1 RxDataRate:1 Quota:20
D (11001) BT_SMP: SMDBG l2c smp_connect_callback

D (11011) BT_SMP: smp_connect_callback()  for pairing BDA: 62a27aca9037  Event: connected

D (11021) BT_SMP: main smp_sm_event

D (11021) BT_SMP: SMP Role: Slave State: [SMP_STATE_IDLE (0)], Event: [L2CAP_CONN_EVT (18)]
D (11031) BT_SMP: State change: SMP_STATE_IDLE(0) ==> SMP_STATE_WAIT_APP_RSP(1)
D (11031) BT_SMP: smp_send_app_cback p_cb->cb_evt=1

D (11041) BT_SMP: io_cap = 4
D (11041) BT_SMP: callback_rc=0  p_cb->cb_evt=1

D (11051) BT_SMP: rcvd auth_req: 0x0d, io_cap: 0                         loc_oob_flag: 0 loc_enc_size: 16,local_i_key: 0x03, local_r_key: 0x03

D (11061) BT_SMP: lmp_version_below LMP version 9 < 8
D (11071) BT_SMP: set auth_req: 0x0d, local_i_key: 0x03, local_r_key: 0x03

D (11071) BT_SMP: main smp_sm_event

D (11081) BT_SMP: SMP Role: Slave State: [SMP_STATE_WAIT_APP_RSP (1)], Event: [API_IO_RSP_EVT (20)]
D (11091) BT_SMP: State change: SMP_STATE_WAIT_APP_RSP(1) ==> SMP_STATE_PAIR_REQ_RSP(3)
D (11091) BT_SMP: smp_process_io_response

D (11101) BT_SMP: State change: SMP_STATE_PAIR_REQ_RSP(3) ==> SMP_STATE_SEC_REQ_PENDING(2)
D (11111) BT_SMP: smp_send_cmd on l2cap cmd_code=0xb

D (11111) BT_SMP: smp_build_security_request

D (11121) BT_SMP: opcode=11 auth_req=0xd
D (11121) BT_SMP: smp_send_msg_to_L2CAP
I (11121) BT_L2CAP: L2CA_SendFixedChnlData()  CID: 0x0006  BDA: 62a27aca9037
D (11131) BT_L2CAP: l2c_link_check_send_pkts
D (11141) BT_L2CAP: partial_segment_being_sent=0,link_state=4,power_mode=0
D (11141) BT_L2CAP: l2c_link_send_to_lower
D (11151) BT_L2CAP: TotalWin=9,Hndl=0x0,Quota=10,Unack=1,RRQuota=0,RRUnack=0
D (11151) BT_SMP: result state = SMP_STATE_SEC_REQ_PENDING

D (11161) BT_SMP: smp_send_app_cback return

D (11161) BT_SMP: result state = SMP_STATE_SEC_REQ_PENDING

D (11171) BT_L2CAP: L2CAP - rcv_cid CID: 0x0006

D (11171) BT_SMP:
SMDBG l2c smp_data_received

D (11181) BT_SMP: main smp_sm_event

D (11181) BT_SMP: SMP Role: Slave State: [SMP_STATE_SEC_REQ_PENDING (2)], Event: [PAIRING_REQ_EVT (1)]
D (11191) BT_SMP: State change: SMP_STATE_SEC_REQ_PENDING(2) ==> SMP_STATE_PAIR_REQ_RSP(3)
D (11201) BT_SMP: smp_proc_pair_cmd

D (11201) BT_SMP: smp_command_has_invalid_parameters for cmd code 0x01

D (11211) BT_SMP: smp_command_has_valid_fixed_length for cmd code 0x01

D (11221) BT_SMP: smp_pairing_request_response_parameters_are_valid for cmd code 0x01

D (11221) BT_HCI: Receive packet event_code=0x13

D (11231) BT_SMP: smp_select_association_model

D (11241) BT_SMP: smp_select_association_model p_cb->peer_io_caps = 4 p_cb->local_io_capability = 0

D (11251) BT_SMP: smp_select_association_model p_cb->peer_oob_flag = 0 p_cb->loc_oob_flag = 0

D (11251) BT_SMP: smp_select_association_model p_cb->peer_auth_req = 0x0d p_cb->loc_auth_req = 0x0d

D (11261) BT_SMP: smp_select_association_model p_cb->secure_connections_only_mode_required = FALSE

D (11271) BT_SMP: use_sc_process = 1

D (11281) BT_SMP: smp_select_association_model_secure_connections

D (11281) BT_SMP: smp_get_auth_mode model 7
D (11291) BT_SMP: smp_send_pair_rsp

D (11291) BT_SMP: smp_send_cmd on l2cap cmd_code=0x2

D (11301) BT_SMP: smp_build_pairing_cmd
D (11301) BT_SMP: smp_send_msg_to_L2CAP
I (11301) BT_L2CAP: L2CA_SendFixedChnlData()  CID: 0x0006  BDA: 62a27aca9037
D (11311) BT_L2CAP: l2c_link_check_send_pkts
D (11321) BT_L2CAP: partial_segment_being_sent=0,link_state=4,power_mode=0
D (11321) BT_L2CAP: l2c_link_send_to_lower
D (11331) BT_L2CAP: TotalWin=8,Hndl=0x0,Quota=10,Unack=2,RRQuota=0,RRUnack=0
D (11331) BT_SMP: smp_decide_association_model Association Model = 7

D (11341) BT_SMP: sec_level=0

D (11341) BT_SMP: main smp_sm_event

D (11351) BT_SMP: SMP Role: Slave State: [SMP_STATE_PAIR_REQ_RSP (3)], Event: [PUBLIC_KEY_EXCHANGE_REQ_EVT (27)]
D (11361) BT_SMP: State change: SMP_STATE_PAIR_REQ_RSP(3) ==> SMP_STATE_PUBLIC_KEY_EXCH(7)
D (11371) BT_SMP: smp_create_private_key
D (11371) BT_HCI: HCI Enqueue Comamnd opcode=0x2018

D (11381) BT_SMP: result state = SMP_STATE_PUBLIC_KEY_EXCH

D (11381) BT_HCI: Receive packet event_code=0xe

D (11391) BT_HCI: btu_hcif_command_complete_evt

D (11391) BT_SMP: result state = SMP_STATE_PUBLIC_KEY_EXCH

D (11401) BT_L2CAP: l2c_link_check_send_pkts
D (11401) BT_L2CAP: partial_segment_being_sent=0,link_state=4,power_mode=0
D (11411) BT_L2CAP: TotalWin=9,LinkUnack(0x0)=1,RRCheck=0,RRUnack=0

D (11411) BT_SMP: smp_rand_back state=0x9
D (11421) BT_SMP: smp_continue_private_key_creation state=0x9

D (11421) BT_HCI: HCI Enqueue Comamnd opcode=0x2018

D (11431) BT_HCI: Receive packet event_code=0xe

D (11431) BT_HCI: btu_hcif_command_complete_evt

D (11441) BT_SMP: smp_rand_back state=0xa
D (11441) BT_SMP: smp_continue_private_key_creation state=0xa

D (11451) BT_HCI: HCI Enqueue Comamnd opcode=0x2018

D (11461) BT_HCI: Receive packet event_code=0xe

D (11461) BT_HCI: btu_hcif_command_complete_evt

D (11461) BT_SMP: smp_rand_back state=0xb
D (11471) BT_SMP: smp_continue_private_key_creation state=0xb

D (11471) BT_HCI: HCI Enqueue Comamnd opcode=0x2018

D (11481) BT_HCI: Receive packet event_code=0xe

D (11481) BT_HCI: btu_hcif_command_complete_evt

D (11491) BT_SMP: smp_rand_back state=0xc
D (11491) BT_SMP: smp_continue_private_key_creation state=0xc

D (11501) BT_SMP: smp_process_private_key
D (11581) BT_SMP: main smp_sm_event

D (11581) BT_SMP: SMP Role: Slave State: [SMP_STATE_PUBLIC_KEY_EXCH (7)], Event: [LOCAL_PUBLIC_KEY_CRTD_EVT (28)]
D (11591) BT_SMP: State change: SMP_STATE_PUBLIC_KEY_EXCH(7) ==> SMP_STATE_PUBLIC_KEY_EXCH(7)
D (11601) BT_SMP: smp_wait_for_both_public_keys

D (11601) BT_SMP: result state = SMP_STATE_PUBLIC_KEY_EXCH

D (16361) BT_HCI: Receive packet event_code=0x13

D (16361) BT_HCI: Receive packet event_code=0x5

D (16361) BT_L2CAP: l2c_link_check_send_pkts
D (16371) BT_L2CAP: partial_segment_being_sent=0,link_state=4,power_mode=0
D (16371) BT_L2CAP: TotalWin=10,LinkUnack(0x0)=0,RRCheck=0,RRUnack=0

D (16381) BT_HCI: btu_free_timer Unable to find expected alarm in hashmap
D (16391) BT_HCI: btu_free_timer Unable to find expected alarm in hashmap
D (16391) BT_HCI: btu_free_timer Unable to find expected alarm in hashmap
D (16401) BT_L2CAP: l2cu_release_ccb: cid 0x0004  in_use: 1
D (16411) BT_HCI: btu_free_timer Unable to find expected alarm in hashmap
D (16411) BT_HCI: btu_free_timer Unable to find expected alarm in hashmap
D (16421) BT_HCI: btu_free_timer Unable to find expected alarm in hashmap
D (16431) BT_L2CAP: l2cu_release_ccb: cid 0x0006  in_use: 1
D (16431) BT_HCI: btu_free_timer Unable to find expected alarm in hashmap
D (16441) BT_SMP: SMDBG l2c smp_connect_callback

D (16441) BT_SMP: smp_connect_callback()  for pairing BDA: 62a27aca9037  Event: disconnected

D (16451) BT_SMP: main smp_sm_event

D (16461) BT_SMP: SMP Role: Slave State: [SMP_STATE_PUBLIC_KEY_EXCH (7)], Event: [L2CAP_DISCONN_EVT (19)]
D (16471) BT_SMP: State change: SMP_STATE_PUBLIC_KEY_EXCH(7) ==> SMP_STATE_IDLE(0)
D (16471) BT_SMP: smp_pair_terminate

D (16481) BT_SMP: smp_proc_pairing_cmpl

D (16481) BT_SMP: send SMP_COMPLT_EVT reason=0x19 sec_level=0x0

D (16491) BT_SMP: smp_reset_control_value

D (16491) BT_SMP: smp_remove_fixed_channel

D (16501) BT_L2CAP: L2CA_RemoveFixedChnl()  CID: 0x0006  BDA: 62a27aca9037 not connected
D (16501) BT_SMP: smp_cb_cleanup

D (16511) BT_SMP: result state = SMP_STATE_IDLE

E (16511) BT_BTM: Device not found

I (16521) BT_L2CAP: L2CA_SetDesireRole() new:x1, disallow_switch:0
I (16521) SEC_GATTS_DEMO: ESP_GATTS_DISCONNECT_EVT, disconnect reason 0x8
I (16541) SEC_GATTS_DEMO: remote BD_ADDR: 62a27aca9037
I (16541) SEC_GATTS_DEMO: address type = 1
I (16541) SEC_GATTS_DEMO: pair status = fail
I (16551) SEC_GATTS_DEMO: fail reason = 0x66
I (16551) SEC_GATTS_DEMO: Bonded devices number : 0
```

## Other items if possible
Elf file for Gatt Demo project:
[sec_gatts_demo_elf.zip](https://github.com/espressif/esp-idf/files/6229736/sec_gatts_demo_elf.zip)

Bluetooth capture for demo project:
[esp_ble_security_capture.zip](https://github.com/espressif/esp-idf/files/6229728/esp_ble_security_capture.zip)

Bluetooth capture for original project:
[ble_stylo6_capture.zip](https://github.com/espressif/esp-idf/files/6229812/ble_stylo6_capture.zip)

**Comment by chewhs00** [comment]
Any feedback on this issue? From the Wireshark capture, after Scan, Feature, and Version req/resp without any problem, then the Master's (mobile phones) Fragment Start packets start to show bad crc and there were no Slave (esp device) responses on all 3 models as specified. The project source code is from 
esp-idf-v4.1.1/examples/bluetooth/bluedroid/ble/gatt_security_server

**Comment by xulongzhuang** [comment]
@rj747 
Could you clear all the bound devices on the phone side first, then erase the flash of the esp32 and re-burn the code, and then perform the pairing, will the same result still appear on those phones?

**Comment by rj747** [comment]
I tested this on multiple new phones that had no prior paired devices and clearing the bound devices was one of the first things I did when I ran into this issue. The attached program was also loaded onto a fresh dev kit as the issue showed up on a custom board with a ESP32. There should be no issues reproducing this with one of the phones listed and we can ship one of them to you if that would help.

**Comment by xiewenxiang** [comment]
@rj747 

The pairing fail is beacause the connection is broken during pairing process. But we can't reproduce this reason by our test phones.

We will purchase these mobile phones.

**Comment by chewhs00** [comment]
@xiewenxiang , any update on this? If necessary, we can provide phones for you to test

**Comment by tomptz** [comment]
@xiewenxiang any update? We are facing a similar problem.

**Comment by xiewenxiang** [comment]
@chewhs00 @tomptz 

This issue has been fixed and will be synchronized out recent.

**Comment by tomptz** [comment]
@xiewenxiang will you inform us in this conversion when the fix is available?

**Comment by xiewenxiang** [comment]
@tomptz The fix has been submitted and is waiting for merged. It is the National Day holiday at present, so it should be after October 7th.

**Comment by tomptz** [comment]
@xiewenxiang any update on this? Is there a development branch that we could test?

**Comment by chewhs00** [comment]
> @chewhs00 did you find a solution or workaround for this issue?

As far as I have tested, Nimble seems to work fine with these phones, but not with Bluedroid

**Comment by AxelLin** [comment]
> @tomptz The fix has been submitted and is waiting for merged.

@xiewenxiang 
Do you mean this fix?
master: a7247db6b4069c7cc0dd67d3cc03b4e03e72c3ed
v4.4: 3187b9b5a4c2eedcfc2be2b83539f74954d9724f

**Comment by xiewenxiang** [comment]
@AxelLin  Yes,  with this commit, can the test pass?

**Comment by chewhs00** [comment]
> > @tomptz The fix has been submitted and is waiting for merged.
> 
> @xiewenxiang Do you mean this fix? master: [a7247db](https://github.com/espressif/esp-idf/commit/a7247db6b4069c7cc0dd67d3cc03b4e03e72c3ed) v4.4: [3187b9b](https://github.com/espressif/esp-idf/commit/3187b9b5a4c2eedcfc2be2b83539f74954d9724f)

The fix solves the pairing issue for LG Stylo 6, LG Phoenix 5, and Alcatel 3V.
The fix currently is in release/v4.1 as well . Any scheduled date for the release?

**Comment by chewhs00** [comment]
Has this fix applied to v4.3.2? Is this the one?

**Comment by AxelLin** [comment]
@xiewenxiang 
v4.3 branch does not include this fix, please backport the fix.
This was reported https://github.com/espressif/esp-idf/issues/8532#issuecomment-1551819583
